Implement X509Certificate2Collection.Find for Unix.

bartonjs opened the pull request:
* All of the object->rich type parsing code is in the common FindPal class.
* Each of the X509FindType types dispatches pre-parsed data to an IFindPal instance.
* There are still two sources of disabled tests for the Unix port
    * Issue 1985: Distingushed Names have (e.g.) "S=Washington" on Windows, "ST=Washington" on Unix.
    * Issue 3099: The Unix and Windows implementations are choosing different certificates out of a multi-file PFX
* Across the various [Theory] components, a little over 100 test cases were added to verify consistent behavior

bartonjs commented:
The tests claimed to be disabled as a part of #1993, but that really tracks `NotImplementedException`s, so there's no issue paperwork to really deal with here.

bartonjs commented:
cc: @AtsushiKan

justinvp commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs:
```diff
@@ -0,0 +1,275 @@
+
+            if (findValueString != null)
+            {
+                if (findValueString.Equals("DigitalSignature", StringComparison.OrdinalIgnoreCase))
```
Could these mappings be in a static readonly Dictionary for faster lookups? (I suspect the Dictionary lookup would be faster than all the if checks, especially for strings towards the end of the if checks, but haven't measured to confirm).

justinvp replied:
Or a `switch` on `findValueString.Length` would be really fast (though, it's more verbose, so might not be worth it depending on how often this code is called)...

Something like:

```c#
switch (findValueString.Length)
{
    case 16:
        switch (char.ToUpperInvariant(findValueString[1]))
        {
            case 'I':
                return ConfirmedX509KeyUsage(findValueString, "DigitalSignature", X509KeyUsageFlags.DigitalSignature);

            case 'A':
                return ConfirmedX509KeyUsage(findValueString, "DataEncipherment", X509KeyUsageFlags.DataEncipherment);
        }
        break;

    case 15:
        return ConfirmedX509KeyUsage(findValueString, "KeyEncipherment", X509KeyUsageFlags.KeyEncipherment);

    case 14:
        return ConfirmedX509KeyUsage(findValueString, "NonRepudiation", X509KeyUsageFlags.NonRepudiation);

    case 12:
        switch (char.ToUpperInvariant(findValueString[0]))
        {
            case 'K':
                return ConfirmedX509KeyUsage(findValueString, "KeyAgreement", X509KeyUsageFlags.KeyAgreement);

            case 'E':
                return ConfirmedX509KeyUsage(findValueString, "EncipherOnly", X509KeyUsageFlags.EncipherOnly);

            case 'D':
                return ConfirmedX509KeyUsage(findValueString, "DecipherOnly", X509KeyUsageFlags.DecipherOnly);
        }
        break;

    case 11:
        return ConfirmedX509KeyUsage(findValueString, "KeyCertSign", X509KeyUsageFlags.KeyCertSign);

    case 7:
        return ConfirmedX509KeyUsage(findValueString, "CrlSign", X509KeyUsageFlags.CrlSign);
}
```

With a helper like this:

```c#
private static X509KeyUsageFlags ConfirmedX509KeyUsage(string findValueString, string flagString, X509KeyUsageFlags flags)
{
    Debug.Assert(findValueString != null);
    Debug.Assert(flagString != null);

    if (flagString.Equals(findValueString, StringComparison.OrdinalIgnoreCase))
    {
        return flags;
    }

    throw new CryptographicException(SR.Cryptography_X509_InvalidFindValue);
}
```

Perhaps with more `Debug.Asserts` to verify the lengths and passed in `flagString` and `flags` match.

bartonjs replied:
While I agree that the jump table is probably a lot faster, dictionary's probably sufficient :)

AtsushiKan commented on src/System.Security.Cryptography.X509Certificates/tests/FindTests.cs:
```diff
@@ -84,9 +86,13 @@ private static void RunSingleMatchTest_PfxCer(X509FindType findType, object find
-                // Verify that the find result and original are linked, not just equal.
-                match.FriendlyName = "HAHA";
-                Assert.Equal("HAHA", expected.FriendlyName);
+                // FriendlyName is Windows-only.
```
Could we use some other property to test the linkage, then?

bartonjs replied:
Nope, FriendlyName and Archived are the only mutatable properties on Windows, and there aren't any on non-Windows.  So there's not really a notion of being linked that needs to be preserved on Unix.

stephentoub commented on src/Common/src/Interop/Unix/libcrypto/Interop.ASN1.Print.cs:
```diff
@@ -0,0 +1,123 @@
+            Asn1StringTypeFlags flags=AnyTextStringType)
+        {
+            SafeAsn1StringHandle asn1String = OpenSslD2I(
+                (zero, ppin, len) => d2i_ASN1_type_bytes(zero, ppin, len, flags),
```
This is still going to create a closure and allocate a delegate due to ```flags```... is that intentional?  This might not be a critical path we need to worry about, so up to you.

bartonjs replied:
It there a way that wouldn't, without inlining the code from OpenSslD2I?

stephentoub replied:
You'd need to change the signature of OpenSslD2I, such that both OpenSslD2I and the delegate provided to it both take an additional state argument... then you can pass flags as that state argument and have it threaded through to the delegate invocation.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs:
```diff
@@ -0,0 +1,268 @@
+            if (maybeOid.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_InvalidOidValue);
+            }
```
Just verifying that ArgumentException is the right exception type for this situation?  Is the OID actually coming in through some public API, and it's the argument to that method we're saying is invalid?

bartonjs replied:
Yep.  `certs.Find(X509FindType.FindByExtension, "", false);`

stephentoub replied:
Ok, cool.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs:
```diff
@@ -0,0 +1,268 @@
+            for (int i = 1; i < len; i++)
+            {
+                // ensure every character is either a digit or a dot
+                if (char.IsDigit(keyValue[i]))
```
IsDigit returns true for characters besides '0'-'9'... is that ok?

bartonjs replied:
Probably not what the original author intended, but it's the behavior desktop has. (http://referencesource.microsoft.com/#System/security/system/security/cryptography/x509/x509utils.cs,6472ac0799e53d68).  I'll have to try a non-ascii value and see if it matches on desktop, I suppose.

bartonjs replied:
\u0662 (arabic 2) doesn't work instead of '2' on Windows, or Unix, returning the empty set instead of complaining the OID is invalid.  Adding a compatibility test, and just leaving this as matching Desktop's behavior.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs:
```diff
@@ -0,0 +1,268 @@
+            // To prevent the big integer from misinterpreted as a negative number,
+            // add a "leading 0" to the byte array.
+            // Since BigInteger(bytes[]) requires a little-endian byte array,
+            // the "leading 0" actually goes at the end of the array.
```
Why isn't this done when the last character is < 0x80?

bartonjs replied:
Because it's already going to be interpreted as positive.  Just ask you asked for back in June :) https://github.com/dotnet/corefx/pull/1990/files#r32145804

stephentoub replied:
Cool; in that case, can you add that to the comment?  The 0x80 looks magical.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs:
```diff
@@ -0,0 +1,268 @@
+                {
+                    accum = BigInteger.Multiply(accum, ten);
+                    accum = BigInteger.Add(accum, c - '0');
+                }
```
else?

bartonjs replied:
Else ignored, sadly.  It's one of the desktop weird input compatibility tests.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslCertificateFinder.cs:
```diff
@@ -0,0 +1,331 @@
+
+        public void FindByThumbprint(byte[] thumbprint)
+        {
+            FindCore(cert => cert.GetCertHash().ContentsEqual(thumbprint));
```
Again, may not matter at all, but this is going to allocate a closure/delegate due to capturing thumbprint.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs:
```diff
@@ -0,0 +1,405 @@
+        public void FindBySubjectDistinguishedName(string subjectDistinguishedName)
+        {
+            FindCore(
+                delegate(SafeCertContextHandle pCertContext)
```
Nit: Why the anonymous method syntax here when you used the lambda syntax elsewhere?

bartonjs replied:
> Why the anonymous method syntax here when you used the lambda syntax elsewhere

These were just moved from the file where they were to this new location during the PAL restructuring.  They were declared delegate in the desktop code, probably because they predated lambdas by about 3 major releases :).

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs:
```diff
@@ -0,0 +1,405 @@
+                    // Convert to BigInteger as the comparison must not fail due to spurious leading zeros
+                    BigInteger actualAsBigInteger = PositiveBigIntegerFromByteArray(actual);
+
+                    GC.KeepAlive(pCertContext);
```
Why the KeepAlive?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs:
```diff
@@ -0,0 +1,405 @@
+
+                    bool foundMatch = false;
+                    CERT_INFO* pCertInfo = pCertContext.CertContext->pCertInfo;
+                    {
```
Why the extra scope?

bartonjs replied:
> Why the extra scope?

Looks like it's just showing the V1 case as a self-contained block.  I just moved it :).

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs:
```diff
@@ -0,0 +1,405 @@
+                    if (numOids == -1)
+                        return true;
+
+                    fixed (byte* pOidsPointer = new byte[cbData])
```
Is there any chance cbData here is 0?

bartonjs replied:
In context, I feel like 'no'.  If there were no valid usages then numOids would have been -1, otherwise there would be non-zero space required for the OID extraction.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs:
```diff
@@ -0,0 +1,405 @@
+
+        private void FindCore(Func<SafeCertContextHandle, bool> filter)
+        {
+            unsafe
```
Nit: You have it in the method signature in other FindCore methods... why not here, too?

bartonjs replied:
> You have [unsafe] in the method signature in other FindCore methods... why not here, too?

The other methods take a void* parameter, so they require it.  By not putting `unsafe` in the declaration here it allows the callers using this one to not also need to say unsafe. 

At least, that's the motivation I assume, since I just moved the existing code to a new file :)

stephentoub replied:
> By not putting unsafe in the declaration here it allows the callers using this one to not also need to say unsafe.

unsafe in the declaration has no impact on callers; it's functionally identical to unsafe instead being used around the entire body of the method.

bartonjs replied:
... I thought something sounded off when I said that.  Justifying other people's code is hard :).

I've made this change, and a few other of the style comments from the Windows code, in a separate commit; to try to keep "move" vs "change" as separate.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs:
```diff
@@ -0,0 +1,405 @@
+                CertStoreFlags.CERT_STORE_ENUM_ARCHIVED_FLAG | CertStoreFlags.CERT_STORE_CREATE_NEW_FLAG,
+                null);
+            if (findResults.IsInvalid)
+                throw Marshal.GetHRForLastWin32Error().ToCryptographicException(); ;
```
Nit: extra ;

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs:
```diff
@@ -0,0 +1,405 @@
+            {
+                Exception verificationException;
+                bool? verified = chainPal.Verify(X509VerificationFlags.NoFlag, out verificationException);
+                if (!(verified.HasValue && verified.Value))
```
```C#
if (!verified.GetValueOrDefault())
```
?

bartonjs commented:
Pushed some new commits.
* Corrected the typo in the base commit title
* Broke the PositiveBigIntegerFromByteArray memory improvement out to a separate commit, to track move vs-change
* Added @justinvp's suggested initial capacity and assert to @justinvp's suggested dictionary :smile:.
* Made some style changes to the Windows FindPal, left them in a separate commit from the base.

stephentoub commented:
LGTM
